Environments: include other environments, optional include destination

tldahlgren opened the pull request:
Supersedes #42442 
Closes?  #46973 

Support for nested configuration includes already exists in Spack and there is a PR for supporting `git` URLs (https://github.com/spack/spack/pull/51191).

This PR makes additional changes (or not) related to environment-specific including other files:

- [x] Supports a _single_ `include` heading in environment manifest (`spack.yaml`) files
- [x]  Deprecates the `include_concrete` heading _in environment manifest files_ 
- [x] _Retains_ use of `include_concrete` in environment lock (`spack.lock`) files for clarity even though the files are not meant to be human-readable
- [x] Supports inclusion of environment manifest (and lock) files under the `include` heading
- [x] _Retains_ support of `--included-concrete` on the environment command line for specifying `spack.lock` files
- [x] Supports nested `include`s _within_ manifest files
- [x] Allows `git` includes without paths to use default path(s) as follows: `spack.yaml` if present or any supported `<config>.yaml` file(s) at the repository root.

TODO
- [x] Update documentation
~- [ ] Implement BFS includes per source comment~  (@tgamblin Indicated diamond and nested includes likely uncommon and implementation should continue to be deferred)
- [x] Add more unit tests to specifically confirm nested includes within included manifest files
- [x] Remove copying of relative include files; instead, ensure the new environment has the absolute path of original file
- [x] Ensure have a unit test with conditional relative paths to confirm the substitution using conditional include paths 
- [x] Update the `git` include documentation and unit tests to reflect the new default paths support 

Per feedback:
 - [x] clone `git` includes for environments under the environment directory (versus under the standard config includes directory)
 - [x] download `path`+`sha256` includes for environments under the environment (versus under standard config includes directory)
 - [x] add optional include destination (to override the normal defaults)
- [x] update documentation according to feedback
- [x] add automatic updates of the include concrete key to the include key + `spack.lock`
~- [ ] fix 'spack config blame' (no sections) to NOT rely on settings in `{section}.yaml` files but process in the same manner as `spack config blame <section>`~
- [x] integrate #51900 changes once those are merged
- [x] integrate #51896 changes once those are merged

Post rebase(s)
- [X] Fix post-rebase handling of explicit include destination
- [X] Fix handling of included environment definitions and toolchains
- [x] Fix handling of included environment specs/groups (IP)
- [ ] Revisit included environments documentation (to address feedback)

tldahlgren committed:
Environments: include other environments

* config.py: use the proper schema for an included environment file
* Fix order types
* Add nested environments schema
* remove types from _env_create arguments docstrings
* Tighten include-concrete processing
* WIP: test_env_include_env tweaks
* Snapshot: include_concrete->include
* Use the include concrete variable for the key
* Remove preliminary handling of include concretes as part of configuration
* Remove IncludePath's is_config
* Snapshot: mainly distinguish include and include-concrete (lockfile) keys
* Snapshot: cleanup and leverage CONFIG (but broken)
* Cleaned up the include-concrete handling and leveraged spack.yaml merged in CONFIG.
  Unfortunately, fixing the simplified new include spack.yaml test ended up breaking
  28 other test/cmd/env.py tests that were working before the last iteration.
* Comment out environment definitions change that broke 28 test/cmd/env.py tests
* Fix style issues
* Update fish command completion
* Include environment definitions in config; add test_env_include_env_pkgs_def
* Drop unnecessary check
* Restore check to get_env_section
* Snapshot of include-concrete-related docs changes
* included_concrete_envs->included_concrete_env_roots; another pass over user specs
* Remove config:_ALL_SCHEMAS since SECTION_SCHEMAS now contains environment
* test/config.py: add test_config_scope_empty_write
* Snapshot test/cmd/env.py: added docstrings to several new tests; drafted test_env_include_env_includes
* config support for spack:include; use hash of included paths in environment names
* Remove unnecessary fixtures from test_env_include_env* tests
* Docs: refined and expanded environment file inclusion
* Comment: added reminder included file copies offline feedback
* Add test_env_include_env_highest_config and switch to BFS nested include processing
* Ugh! test_env_include_env_includes SHOULD use mutable_config
* test_include_recurse_diamond: Fix to use the required section name for top-most include
* Add missing ref link; fix style
* Restore DFS nested include handling
  Also restore the docs warning and change new BFS unit test to XFAIL.
* support include path rewrites
* Remove commented out copy operations
* test_gitlab_config_scopes: recognize includes of new environment are absolute
* test_gitlab_config_scopes: confirm ignoring duplicate included scopes output
* test_env_include_envs: switch to conditional path syntax to confirm path substitution retains information
* Restore use of string variable for unit tests
  Appears some pythons handle embedding a join but others don't.
* test_env_include_envs: use python@:3.8 removeprefix solution
* Make sure cached included concrete path is canonicalized
* Post-rebase cleanup
* config.py: Fix docstring typos
* environment.py: add debug for adding lockfile to concrete roots
* test/config.py: add test_config_optional_include_failures
* test_config_optional_include_failures: fix style
* Filter out lockfiles from include scopes
* OptionalIncludes: Leverage simplified property checks
* config.push_scope: restore simplified included scope processing
* test/cmd/env.py: add test_env_include_concrete_only
* Support default paths with git includes (per offline feedback)
* Initial rebase fix
* Fish completion update: adds v/verbose
* config.get_config: added debug messages to indicate source
* Bugfix: don't try to update environment data scope; more debug statements
* git includes: clone environments under the environment hidden directory
* Add missing space to debug message
* Initial docs updates from PR feedback
* Cache env path_sha256 includes under the environment.
* config.py: revise the manifest precedence comment
* Rebase cleanup (e.g.,remove unused _ALL_SCHEMAS)
* Address remaining documentation feedback
* Support remote include destination:
* Optional remote include path destination
* Fix a couple of remote include unit tests' sha256s
* test_included_path_git_default_paths: Add optional explicit remote include destination
* Ensure variable substitution remote include destination
* Docs: update include.yaml with 'destination' option
* 'spack env update' convert include_concrete to include
* Docs: Add section on 'spack env update'
* Remove schema/env.py -> spack.environment circular import
* Added test_include_concrete_deprecation_warning and test_env_schema_update_wrong_type
* config: try canonicalize through rfc_util
* Check if include destination is writable; add unit tests
* Disable unwritable directories for windows
* Flag new 'destination:' option as optional
* test/cmd/emv/py: Rebase imports cleanup
* (WIP) _sync_speclists: comment out the rest of the inclusion
* Post-rebase style fixes
* Post-rebase schema fix
* test/cmd/env.py: e.all_concretized_user_specs -> e.user_specs
* Post-rebase cleanup
* Add included toolchain (needs tests)

Signed-off-by: tldahlgren <dahlgren1@llnl.gov>

tldahlgren committed:
Environments: include other environments

* config.py: use the proper schema for an included environment file
* Fix order types
* Add nested environments schema
* remove types from _env_create arguments docstrings
* Tighten include-concrete processing
* WIP: test_env_include_env tweaks
* Snapshot: include_concrete->include
* Use the include concrete variable for the key
* Remove preliminary handling of include concretes as part of configuration
* Remove IncludePath's is_config
* Snapshot: mainly distinguish include and include-concrete (lockfile) keys
* Snapshot: cleanup and leverage CONFIG (but broken)
* Cleaned up the include-concrete handling and leveraged spack.yaml merged in CONFIG.
  Unfortunately, fixing the simplified new include spack.yaml test ended up breaking
  28 other test/cmd/env.py tests that were working before the last iteration.
* Comment out environment definitions change that broke 28 test/cmd/env.py tests
* Fix style issues
* Update fish command completion
* Include environment definitions in config; add test_env_include_env_pkgs_def
* Drop unnecessary check
* Restore check to get_env_section
* Snapshot of include-concrete-related docs changes
* included_concrete_envs->included_concrete_env_roots; another pass over user specs
* Remove config:_ALL_SCHEMAS since SECTION_SCHEMAS now contains environment
* test/config.py: add test_config_scope_empty_write
* Snapshot test/cmd/env.py: added docstrings to several new tests; drafted test_env_include_env_includes
* config support for spack:include; use hash of included paths in environment names
* Remove unnecessary fixtures from test_env_include_env* tests
* Docs: refined and expanded environment file inclusion
* Comment: added reminder included file copies offline feedback
* Add test_env_include_env_highest_config and switch to BFS nested include processing
* Ugh! test_env_include_env_includes SHOULD use mutable_config
* test_include_recurse_diamond: Fix to use the required section name for top-most include
* Add missing ref link; fix style
* Restore DFS nested include handling
  Also restore the docs warning and change new BFS unit test to XFAIL.
* support include path rewrites
* Remove commented out copy operations
* test_gitlab_config_scopes: recognize includes of new environment are absolute
* test_gitlab_config_scopes: confirm ignoring duplicate included scopes output
* test_env_include_envs: switch to conditional path syntax to confirm path substitution retains information
* Restore use of string variable for unit tests
  Appears some pythons handle embedding a join but others don't.
* test_env_include_envs: use python@:3.8 removeprefix solution
* Make sure cached included concrete path is canonicalized
* Post-rebase cleanup
* config.py: Fix docstring typos
* environment.py: add debug for adding lockfile to concrete roots
* test/config.py: add test_config_optional_include_failures
* test_config_optional_include_failures: fix style
* Filter out lockfiles from include scopes
* OptionalIncludes: Leverage simplified property checks
* config.push_scope: restore simplified included scope processing
* test/cmd/env.py: add test_env_include_concrete_only
* Support default paths with git includes (per offline feedback)
* Initial rebase fix
* Fish completion update: adds v/verbose
* config.get_config: added debug messages to indicate source
* Bugfix: don't try to update environment data scope; more debug statements
* git includes: clone environments under the environment hidden directory
* Add missing space to debug message
* Initial docs updates from PR feedback
* Cache env path_sha256 includes under the environment.
* config.py: revise the manifest precedence comment
* Rebase cleanup (e.g.,remove unused _ALL_SCHEMAS)
* Address remaining documentation feedback
* Support remote include destination:
* Optional remote include path destination
* Fix a couple of remote include unit tests' sha256s
* test_included_path_git_default_paths: Add optional explicit remote include destination
* Ensure variable substitution remote include destination
* Docs: update include.yaml with 'destination' option
* 'spack env update' convert include_concrete to include
* Docs: Add section on 'spack env update'
* Remove schema/env.py -> spack.environment circular import
* Added test_include_concrete_deprecation_warning and test_env_schema_update_wrong_type
* config: try canonicalize through rfc_util
* Check if include destination is writable; add unit tests
* Disable unwritable directories for windows
* Flag new 'destination:' option as optional
* test/cmd/emv/py: Rebase imports cleanup
* (WIP) _sync_speclists: comment out the rest of the inclusion
* Post-rebase style fixes
* Post-rebase schema fix
* test/cmd/env.py: e.all_concretized_user_specs -> e.user_specs
* Post-rebase cleanup
* Add included toolchain (needs tests)

Signed-off-by: tldahlgren <dahlgren1@llnl.gov>
Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

tgamblin commented on lib/spack/docs/environments.rst:
```diff
@@ -265,18 +265,18 @@ In the :ref:`environment-configuration` section we will discuss environment-awar
-Adding a spec adds it as a root spec of the environment in the user input file (``spack.yaml``).
-It does not affect the concrete specs in the lock file (``spack.lock``) and it does not install the spec.
+The addition of abstract specs does not affect the concrete specs in the lock file (``spack.lock``) nor results in the automatic installation of the added specs.
+Instead, the environment must be :ref:`re-concretized <cmd-spack-concretize>` or :ref:`(re)installed <installing-environment>`.
```
```suggestion
Adding an abstract spec does not affect concrete specs in the lock file (``spack.lock``), and it does not automatically install the added specs.
Instead, the environment must be :ref:`re-concretized <cmd-spack-concretize>` or :ref:`(re)installed <installing-environment>`.
```

tgamblin replied:
```suggestion
Adding an abstract specs does not immediately install anything, nor does it affect the ``spack.lock`` file.
To update the lockfile, the environment must be :ref:`re-concretized <cmd-spack-concretize>`, and to update any installations, the environment must be :ref:`(re)installed <installing-environment>`.
```

tgamblin commented on lib/spack/docs/environments.rst:
```diff
@@ -586,16 +484,20 @@ For example, a ``spack.yaml`` manifest file containing some package preference c
 Spack environments allow an ``include`` heading in their yaml schema.
-This heading pulls in external configuration files and applies them to the environment.
+This heading pulls in external configuration **and** environment files and applies them to the environment.
+(See :ref:`environment_inclusion` for more information on including environment files.)
```
```suggestion
`include:` pulls in external configuration (including other environments) and applies them to the environment.
(See :ref:`environment_inclusion` for more information on including environment files.)
```

tgamblin commented on lib/spack/spack/environment/environment.py:
```diff
@@ -1081,43 +1033,74 @@ def add_view(name, values):
+            tty.warn(
+                f"Use of '{lockfile_include_key}' in manifest files "
+                f"is deprecated. Please change the key to '{manifest_include_name}'."
+                f" And make sure to end each path with '{lockfile_name}'."
```
probably needs an update() method in config schema so this can be done automatically with `spack config update`

tldahlgren replied:
I think you mean `spack env update <managed-name | independent-path>`.

tldahlgren replied:
I've added support and documentation for this. Unfortunately, I also introduced a circular import when I decided it was important to support path variable substitutions for the destination of an include.

tldahlgren replied:
Added a "workaround" to avoid the new circular import by using `canonicalize_path`, which is what is consistent with the API used against the `repos` destination.

tldahlgren commented on lib/spack/spack/environment/environment.py:
```diff
@@ -1081,43 +1033,74 @@ def add_view(name, values):
+                continue
+
+            if os.path.basename(include.path) == lockfile_name:
+                # Assumes the included path is already canonicalized.
```
This assumption is not necessarily accurate so need to ensure it is canonicalized.

tldahlgren commented:
@tgamblin Looks like it's having the same CI problem it was before my rebase:

```
Preparing environment
00:10
Using FF_USE_POD_ACTIVE_DEADLINE_SECONDS, the Pod activeDeadlineSeconds will be set to the job timeout: 6h0m0s...
Waiting for pod pipeline/runner-mhb1or73u-project-2-concurrent-2-fq894hvm to be running, status is Pending
Waiting for pod pipeline/runner-mhb1or73u-project-2-concurrent-2-fq894hvm to be running, status is Pending
	ContainersNotInitialized: "containers with incomplete status: [init-permissions]"
	ContainersNotReady: "containers with unready status: [build helper]"
	ContainersNotReady: "containers with unready status: [build helper]"
Waiting for pod pipeline/runner-mhb1or73u-project-2-concurrent-2-fq894hvm to be running, status is Pending
	ContainersNotReady: "containers with unready status: [build helper]"
	ContainersNotReady: "containers with unready status: [build helper]"
Running on runner-mhb1or73u-project-2-concurrent-2-fq894h
```

tgamblin commented on lib/spack/docs/environments.rst:
```diff
@@ -265,18 +265,18 @@ In the :ref:`environment-configuration` section we will discuss environment-awar
 Adding Abstract Specs
 ^^^^^^^^^^^^^^^^^^^^^
 
+The most important component of an environment is its list of abstract specs.
```
I think this isn't necessarily true. An environment (especially after this PR) may just be some config that someone wants to share.

tldahlgren replied:
Fair point though if there are no abstract specs associated with the environment then there is nothing to build, right?

tgamblin replied:
The point is that I could do this in my own environment:

```yaml
spack:
  include:
    - git: https://github.com/llnl/lc-spack
      commit: 9be36efacf1a5fe750b35758030f3ecccf0aa0da
      paths:
        - elcap/rocm-6.1
  specs:
    - py-torch
    - myapp
    - tool1
    - tool2
```

And my specs would build with whatever LC's preferred settings for that ROCm on that machine are, and the user does not have to set anything up.

I would argue that the most important part of the (currently fictitious) `lc-spack` environments is their configuration, which allows others not to have to duplicate it or even think about it.  They eventually provide specs, but in the including environment, not the included one.

tldahlgren replied:
I did get your point. Thanks for the extra clarification.

tgamblin commented on lib/spack/docs/environments.rst:
```diff
@@ -265,18 +265,18 @@ In the :ref:`environment-configuration` section we will discuss environment-awar
-Users can add abstract specs to an environment using the ``spack add`` command.
-The most important component of an environment is a list of abstract specs.
+Users can add abstract specs to an environment using the ``spack add`` command and.
+The result is the spec is added as a root spec of the environment in the user input, or manifest, file (``spack.yaml``).
```
```suggestion
You can add abstract specs to an environment with the ``spack add`` command, which adds each abstract spec as a root of the environment in the ``spack.yaml`` file.
```

tgamblin commented on lib/spack/docs/environments.rst:
```diff
@@ -628,6 +530,120 @@ Configuration precedence
+Spack environments can build on other environments.
+You can think of them as combinations of existing environments.
+A combined environment is one in which Spack starts with the base environment's configuration and integrates information from existing environments that it includes.
+Existing environments may be specified using their abstract or concrete files.
```
```suggestion
Existing environments may be specified using their abstract (`spack.yaml`) or concrete (`spack.lock`) files.
```

tgamblin commented on lib/spack/docs/environments.rst:
```diff
@@ -628,6 +530,120 @@ Configuration precedence
+Existing environments may be specified using their abstract or concrete files.
+
+Including an abstract environment involves adding its ``spack.yaml`` file to the new environment.
+This results in the integration of the configuration information from the included environment into that of the combined environment.
```
I think this should be expanded on a bit with examples

tgamblin commented on lib/spack/docs/environments.rst:
```diff
@@ -628,6 +530,120 @@ Configuration precedence
+(See :ref:`included_configurations` for more information.)
+Concrete environment inclusion is different in that the integration involves copying a snapshot of the included concrete environment into the combined concrete environment.
+
+The remainder of this section focuses on including concrete environments.
```
I feel like this reads like a tech paper -- the documentation should be relatively casual, easy to read, and shouldn't need a roadmap... honestly should make more use of active voice and second person

```suggestion
```

tgamblin commented on lib/spack/docs/environments.rst:
```diff
@@ -628,6 +530,120 @@ Configuration precedence
+Including an abstract environment involves adding its ``spack.yaml`` file to the new environment.
+This results in the integration of the configuration information from the included environment into that of the combined environment.
+
+Including a concrete environment is accomplished by specifying its ``spack.lock`` file.
```
```suggestion
You can include a concrete environment by specifying its ``spack.lock`` file.
```

tgamblin commented on lib/spack/docs/environments.rst:
```diff
@@ -628,6 +530,120 @@ Configuration precedence
+This results in the integration of the configuration information from the included environment into that of the combined environment.
+
+Including a concrete environment is accomplished by specifying its ``spack.lock`` file.
+In this case, the included concrete environment is integrated into the combined environment's ``spack.lock`` file.
```
```suggestion
The included concrete environment is then added to the combined environment's ``spack.lock`` file.
```

tgamblin commented on lib/spack/docs/environments.rst:
```diff
@@ -628,6 +530,120 @@ Configuration precedence
+In this case, the included concrete environment is integrated into the combined environment's ``spack.lock`` file.
+
+Adding abstract environments to an environment is similar to including other configuration files in that the configuration information is integrated in essentially the same manner.
+(See :ref:`included_configurations` for more information.)
```
Needs more elaboration

tgamblin commented on lib/spack/spack/test/env.py:
```diff
@@ -1318,3 +1318,127 @@ def test_reuse_environment_dependencies(tmp_path: pathlib.Path, mutable_config):
+def test_env_include_env_includes(tmp_path, mutable_config):
+    """Confirm that an environment that includes two other environments,
+    each including their own configuration, picks up the specs from both
+    but its own configuration setting."""
```
I don't think I'm understanding this -- an env that includes another should be including the config from the other environment, right?  Not just the specs.  I *believe* the docs above for `environments.rst` imply that.

I tried to make a really basic test case here:

* https://github.com/tgamblin/env-test/blob/main/spack.yaml

To show that I could get `concretizer:unify:true` to show up based on an environment include.  I tried this env:

```yaml
spack:
  include:
    - git: git@github.com:tgamblin/env-test
      branch: main
      paths:
        - spack.yaml
    - ~/src/spack/var/spack/environments/default/spack.lock
```

The concrete include works fine but if I try to look at the remote include, I get:

```console
> spack -e . config blame concretizer
---                                                                          concretizer:
/Users/gamblin2/Workspace/src/spack/etc/spack/defaults/concretizer.yaml:18     reuse: True
/Users/gamblin2/Workspace/src/spack/etc/spack/defaults/concretizer.yaml:23     targets:
/Users/gamblin2/Workspace/src/spack/etc/spack/defaults/concretizer.yaml:28       granularity: microarchitectures
/Users/gamblin2/Workspace/src/spack/etc/spack/defaults/concretizer.yaml:32       host_compatible: True
/Users/gamblin2/Workspace/src/spack/etc/spack/defaults/concretizer.yaml:38     unify: True
...
```

Note that the new env, and its `concretizer:unify:true`, doesn't show up anywhere -- is that intended?  I do not think it should be.

tldahlgren replied:
Thanks for the example. I too see it isn't grabbing the setting from the included `spack.yaml`.

tldahlgren replied:
I found the problem and have fixed it for this particular use case.

The output from `spack config blame` does not YET merge the included setting as expected, but in this case throughs the options (for me) at the end.

tldahlgren replied:
I am no longer having the `spack config blame` issue post-rebase.

tgamblin reviewed (changes requested):
I have some questions around remote environment includes and around the storage model for includes.  I think we should think through what the storage model ought to looks like in and outside of environments. 

I *think* it makes sense to download the included files in an env *into* the env itself. For configuration files I think the current location (`~/.spack/includes`) looks pretty good, though there is no `spack clean` option for them and I wonder if they should either go in `~/.spack/cache` or have their own clean option.  I am not sure how to reconcile the need for `~/.spack` for includes outside an environment, when the env essentially is its own working directory and a logical place for this stuff to go.

See below -- I tried to test this but wasn't able to get it to include a simple config setting from a git repo.

Based on the uncertainty I think it makes sense to move this stuff to 1.2 and to figure out where these files really belong and what the initial build / install / update / upgrade workflow would look like, so moving this to 1.2.

tldahlgren commented:
> > I _think_ it makes sense to download the included files in an env _into_ the env itself. For configuration files I think the current location (`~/.spack/includes`) looks pretty good, though there is no `spack clean` option for them and I wonder if they should either go in `~/.spack/cache` or have their own clean option. I am not sure how to reconcile the need for `~/.spack` for includes outside an environment, when the env essentially is its own working directory and a logical place for this stuff to go.
> 
> Git includes for an environment are now cloned under `<ENV>/.spack-env/includes/<hash>`.

That is not currently the case for an environment including the URL (and sha256) for a raw configuration file (e.g., https://spack.readthedocs.io/en/latest/include_yaml.html#remote-file-urls).

tldahlgren commented:
FYI. CI had 14 failures (https://gitlab.spack.io/spack/spack/-/pipelines/1310253/failures) at commit https://github.com/spack/spack/pull/50207/commits/60562878efdb52a1b72754c486c88ef9d7f39423, one of which was:

```
...
$ spack buildcache update-index --keys s3://spack-binaries-prs/spack/pr50207_env_unified_include/e4s-cray-rhel
==> Warning: Failed to use aws s3 sync to retrieve specs, falling back to parallel fetch
==> Error: Command exited with status 1:
'/usr/local/bin/aws' 's3' 'ls' '--recursive' 's3://spack-binaries-prs/spack/pr50207_env_unified_include/e4s-cray-rhel'

Cleaning up project directory and file based variables
```

alecbcs commented on lib/spack/docs/include_yaml.rst:
```diff
@@ -52,7 +52,21 @@ A ``sha256`` is required and must be specified as follows:
+.. code-block:: yaml
+
+   include:
+   - path: https://github.com/path/to/raw/config/config.yaml
```
```suggestion
   - path: https://raw.githubusercontent.com/path/to/raw/config/config.yaml
```

Unless we're doing anything special with GitHub URLs, I think we might want to update the example to use raw.githubusercontent.com instead of github.com in the example here. (The latter will give you an HTML version of the file viewer instead of the raw file.)

tldahlgren replied:
We are for GitHub *and* GitLab. See https://github.com/spack/spack/blob/044f2902da3600b97a61ae6ce32b3e96940b2351/lib/spack/spack/util/remote_file_cache.py#L21.

tldahlgren replied:
The goal was to make it easy for users to specify the URL by performing the conversion in spack.

alecbcs commented on lib/spack/docs/include_yaml.rst:
```diff
@@ -52,7 +52,21 @@ A ``sha256`` is required and must be specified as follows:
+   include:
+   - path: https://github.com/path/to/raw/config/config.yaml
+     sha256: 26e871804a92cd07bb3d611b31b4156ae93d35b6a6d6e0ef3a67871fcb1d258b
+     destination: $HOME/my/custom/remote/cache/directory
```
```suggestion
     destination: $HOME/my/custom/remote/cache/directory # optional
```

tldahlgren replied:
I was wondering if I wasn't being clear enough in the description. Fixed in two places *and* added to their `versionadded`. Thanks.

alecbcs reviewed (commented):
Two quick questions, otherwise this is looking pretty good to me. This has been a long time coming and I don't think it necessarily needs to be perfect when first merged as we can continue iterating in smaller follow up PRs on some of docs, etc...

For this PR however, what's the method to update an included file in an environment when it's coming from a git repository and a branch? E.g. if the upstream included repository updates how to we propagate those changes down to a user's install. For repositories we do this with `spack repo update` which isn't perfect, but it at least a fairly intuitive way to manage changing repos.

tldahlgren commented:
Looks  like  a container issue in at least one job. Trying again in case it's intermittent.

@spackbot run pipeline

becker33 commented on lib/spack/spack/test/config.py:
```diff
@@ -1692,31 +1696,52 @@ def test_included_path_conditional_bad_when(
+        entry["sha256"] = "1e44af11d28bb1ddce335ff90129d6b081faf067930f1bb9b745990c6958d8b6"
+
+    include = spack.config.included_path(entry)
+    assert include.scopes(mock_low_high_config.scopes["low"])
```
Can you separate this into two tests, one that tests the error for missing sha256 and one that tests the destination?

becker33 commented on lib/spack/spack/test/config.py:
```diff
@@ -1960,3 +2043,71 @@ def test_missing_include_scope_write_file(mock_missing_file_include_scopes):
+        with pytest.raises(AssertionError, match="to unwritable"):
+            include.scopes(parent_scope)
+    finally:
+        dest.chmod(current_mode)
```
```suggestion
    with pytest.raises(AssertionError, match="to unwritable"):
        include.scopes(parent_scope)
```
No need to replace the mode on a temp file

becker33 commented on lib/spack/docs/include_yaml.rst:
```diff
@@ -52,7 +52,21 @@ A ``sha256`` is required and must be specified as follows:
 
-The ``config.yaml`` file would be cached locally to a special include location and its contents included in Spack's configuration.
+The ``config.yaml`` file would be cached locally and its contents included in Spack's configuration.
+You can add a ``destination`` to the entry to specify the desired cache directory; otherwise, a special default include location will be used.
```
Please explain where the default location is instead of saying that it's "special".

becker33 commented on lib/spack/docs/environments.rst:
```diff
@@ -452,79 +454,207 @@ Sourcing that file in Bash will make the environment available to the user, and
+
+Spack environments can build on other environments.
+You can think of them as combinations of existing environments.
+A combined environment is one in which Spack starts with the enclosing environment's configuration then adds configuration the from the existing environments that it includes.
```
```suggestion
A Spack environment consists of configuration files, and like any other configuration files can be included in another environment.
```
I think the language about combination here is distracting from the main point, making it sound like this behaves differently from other includes.

alalazo replied:
Typo:
```suggestion
A combined environment is one in which Spack starts with the enclosing environment's configuration then adds configuration from the existing environments that it includes.
```
Also, the sentence as is is not super clear to me 🤔 

Maybe we can start from a simple motivating use case, and show how including a remote environment allows to have a shared "base configuration" among multiple users, that can be overridden by the enclosing environment.

becker33 commented on lib/spack/docs/environments.rst:
```diff
@@ -452,79 +454,207 @@ Sourcing that file in Bash will make the environment available to the user, and
+URLs to individual files must link to the **raw** form of the file's contents (e.g., `GitHub <https://docs.github.com/en/repositories/working-with-files/using-files/viewing-and-understanding-files#viewing-or-copying-the-raw-file-content>`_ or `GitLab <https://docs.gitlab.com/ee/api/repository_files.html#get-raw-file-from-repository>`_) **and** include a valid sha256 for the file.
+Only the ``file``, ``ftp``, ``http`` and ``https`` protocols (or schemes) are supported.
+Spack-specific, environment and user path variables can be used.
+(See :ref:`config-file-variables` for more information.)
```
This paragraph needs to be reworked to clarify that includes can come from git repos in addition to paths/urls.

becker33 commented on lib/spack/docs/environments.rst:
```diff
@@ -452,79 +454,207 @@ Sourcing that file in Bash will make the environment available to the user, and
-If changes were made to the base environment and you want that reflected in the included environment you will need to re-concretize both the base environment and the included environment for the change to be implemented.
+Updating a combined concrete environment
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+If changes are made to the included environment that you want reflected in the combined environment, you will need to re-concretize both the included environment then the combined environment for the changes to be integrated.
```
```suggestion
Updating an environment that includes concrete specs from another environment
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Concrete specs included from another environment are updated based on the included environment's state at concretization time of the including environment.
To incorporate changes from the manifest of the included environment, first concretize the included environment and then reconcretize the including environment to integrate the changes.
```
I don't think the "combined" environment vocabulary was helping us here.

becker33 commented on lib/spack/spack/config.py:
```diff
@@ -801,8 +823,20 @@ def _get_config_memoized(
         # filter any scopes overridden by `include::`
         scopes = self._filter_overridden(scopes)
 
+        def get_env_section(config_scope, section):
+            data = config_scope.get_section(spack.schema.env.TOP_LEVEL_KEY)
+            if data is None:
+                return None
+
+            # Grab the actual data
+            if spack.schema.env.TOP_LEVEL_KEY in data:
+                data = data[spack.schema.env.TOP_LEVEL_KEY]
+
+            return data
```
I think this logic should also be pushed into `get_section`

alalazo replied:
In this function we never use the `section` argument, so I suspect there's a bug. I still have to fully understand the logic we use to recursively import included scopes though 😟

alalazo commented on lib/spack/docs/environments.rst:
```diff
@@ -452,79 +454,207 @@ Sourcing that file in Bash will make the environment available to the user, and
+
+Spack environments allow an ``include`` heading in their yaml schema.
+This heading pulls in external configuration **and** environment files and applies them to the environment.
+(See :ref:`environment_inclusion` for more information on including environment files.)
```
(Comment, not necessarily for this PR) There is quite a bit of redundancy between this part and the `includes.yaml` section. Wondering if we can just refer to the other section to explain the mechanics of `include:`.

alalazo commented on lib/spack/docs/environments.rst:
```diff
@@ -452,79 +454,207 @@ Sourcing that file in Bash will make the environment available to the user, and
+   If you include a remote repository or local directory, the abstract environment takes precedence.
+   In other words, Spack will choose the ``spack.yaml`` file over all other configuration files by default.
+   If you want other configuration files included, you will need to explicitly specify them.
+   If you want to include a concrete environment, you will similarly need to explicitly include its ``spack.lock`` file.
```
Instead of this being a warning, can we move it in the `include_yaml.rst` paragraph as a brief section?

alalazo commented on lib/spack/docs/environments.rst:
(Just a suggestion, which could be discussed as possibly controversial. Can be done in a following PR) I would try to keep:
1. Including a remote `spack.yaml` into the local `spack.yaml`
2. Including a concrete environment into the local environment

two separate sections, instead of being two sub-sections of the "Including Environments" section. The rationale is that the use cases are very different, so we might introduce them better if they are separate from each other.

alalazo commented on lib/spack/docs/include_yaml.rst:
```diff
@@ -78,12 +94,26 @@ We would then configure the ``include.yaml`` file as follows:
+
+.. note::
+
+   If a list of relative paths to the configuration files is not provided, the root directory of the repository will be checked for suitable files.
```
Is this behavior a current limitation, or do we want to keep this behavior also in the long run? If the latter, I would take it out of a `Note`.

alalazo commented on lib/spack/spack/cmd/env.py:
```diff
@@ -558,8 +558,8 @@ def _env_untrack_or_remove(
-            # check if an environment is included un another
-            if remove_env.path in env.included_concrete_envs:
+            # check if an environment is included in another
+            if remove_env.path in env.included_concrete_env_roots:
```
Besides this line, changes here are "trivial" fixes. If you want to extract these kind of changes (including similar ones in other files) into another PR we can merge them immediately.

tldahlgren replied:
See #51899

alalazo commented on lib/spack/spack/environment/environment.py:
```diff
@@ -963,6 +904,17 @@ def env_subdir_path(manifest_dir: Union[str, pathlib.Path]) -> str:
+      default: optional default value if key is not detected under ``spack:``
+    """
+    data = spack.config.CONFIG.get(TOP_LEVEL_KEY, {})
+    return data.get(key, default)
```
I don't understand why this function talks about the "included" environment in the docstring, while it pulls values out of the global `CONFIG`

tldahlgren replied:
If the environment includes other environments, they will show up in the global `CONFIG` under the key `spack`.

alalazo commented on lib/spack/spack/environment/environment.py:
```diff
@@ -996,8 +948,8 @@ def __init__(self, manifest_dir: Union[str, pathlib.Path]) -> None:
-        #: Environment paths for concrete (lockfile) included environments
-        self.included_concrete_envs: List[str] = []
+        #: Environment roots for concrete included environments
+        self.included_concrete_env_roots: List[str] = []
```
Can we call these:
```suggestion
        #: Environment root dirs for concrete included environments
        self.included_concrete_env_root_dirs: List[str] = []
```
(or anything similar) to disambiguate with the term "roots" used as in the user specs listed under `specs:` in the included environment?[^1]

[^1]: That would be the case for e.g. the comment on `concretized_order` above

tldahlgren replied:
See #51899

alalazo commented on lib/spack/spack/config.py:
```diff
@@ -148,29 +145,54 @@ def __init__(self, name: str) -> None:
+    def get_includes(self) -> Optional[YamlConfigDict]:
+        """Retrieve the 'include' section, allowing for the fact that includes
+        nested in a manifest file fall under spack:include."""
+        includes = self.get_section("include")
+        if includes:
+            tty.debug(f"Retrieved configuration includes for {self.name}: {includes}", level=3)
+            return includes
```
Not clear to me at the moment why we search first for `include:` and then for `spack:include:` and not the other way around.

alalazo reviewed (commented):
I had an initial read of most of the the PR (unit tests still missing), and left a few comments. At a high-level my suggestions would be:
1. Let's try to extract anything that is a refactoring in preparation for the feature added in this PR to other PRs, to be merged quickly. This may reduce by the amount of lines for this PR by 1-2 hundreds.
2. I'd keep the documentation a bit more succint (but that is also  a personal preference, so feel free to disregard)

I'll continue later with unit tests and test driving of the PR. Will post here any further comment.

alalazo commented on lib/spack/spack/test/cmd/ci.py:
```diff
@@ -1593,6 +1594,8 @@ def test_gitlab_config_scopes(install_mockery, ci_generate_test, tmp_path: pathl
 """
     )
+    _, err = capfd.readouterr()
+    assert err.count("Ignoring duplicate included scope") > 2
```
(This is more for me to understand the test, than anything else) Do you know why we construct the test here including multiple times the same environment in slightly different ways:
```yaml
include:
  - gitlab/configs
  - path: gitlab/configs
  - /tmp/pytest-of-culpo/pytest-11/test_gitlab_config_scopes0/gitlab/configs
```
? What is this meant to test? Would it be better to parametrize the test and use one include at a time?

alalazo commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -1076,12 +1075,7 @@ def test_init_from_env_no_spackfile(tmp_path):
-        "repos.yaml",
-        "./config.yaml",
-    ]
+    files = ["relative/packages.yaml", "./relative/compilers.yaml", "repos.yaml", "./config.yaml"]
```
(Not an issue introduced in this PR) Here we can probably add a docstring, explaining the use case being tested:

```suggestion
    """Tests that we can create an environment starting from a manifest file with relative include paths.
    
    Relative paths are turned into absolute paths in the new manifest.
    """
    files = ["relative/packages.yaml", "./relative/compilers.yaml", "repos.yaml", "./config.yaml"]
```

alalazo commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -1099,14 +1093,19 @@ def test_init_from_yaml_relative_includes(tmp_path: pathlib.Path):
+        assert new.endswith(orig.strip("./"))
+
+    # Check configured environment's includes
+    config_includes = ev.included_env_config("include", [])
```
This is actually calling the global configuration, and since the environment is not active, we get an empty list (and the test passes)

<img width="1780" height="251" alt="Image" src="https://github.com/user-attachments/assets/cc8203f3-3c79-4bb8-8578-fc0e8e0afede" />

alalazo commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -1099,14 +1093,19 @@ def test_init_from_yaml_relative_includes(tmp_path: pathlib.Path):
+    # Check new environment's (manifest) includes
+    new_env_includes = e2.manifest.configuration.get("include", [])
+    for orig, new in zip(files, new_env_includes):
+        assert new.endswith(orig.strip("./"))
```
Is it worth checking something similar to:
```python
assert new == str((e1_path / orig).resolve())
```
? Would it always be true that we'll get absolute paths in the new environment?

alalazo commented on lib/spack/spack/test/cmd/env.py:
(Just a suggestion) If we introduce `manifest_include_name` and `lockfile_include_key` in a separate PR, I think we can probably extract many trivial changes from this file.

tldahlgren replied:
At this point, off `develop`, I can do `lockfile_include_key` independently. Changes here related to using `manifest_include_name` are part of changes for unifying includes under a single `include` key.

tldahlgren replied:
See #51899 for the `lockfile_include_key` change. The changes for `manifest_include_name` will need to go into a follow-on PR since they're part of unifying includes.

alalazo commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -4635,3 +4633,226 @@ def test_non_str_repos(installed_environment):
+    prefix = str(tmp_path) + os.sep
+    # TODO: Once minimum python is python@3.9
+    # rel_paths = [include.removeprefix(prefix) for include in includes]
+    rel_paths = [remove_prefix(include, prefix) for include in includes]
+    paths_str = "\n  ".join([include_entry(path) for path in rel_paths])
+    e = environment_from_manifest(
+        f"""\
+spack:
+  include:
+  {paths_str}
+"""
```
If we are more explicit about the `paths_str`, we'll document better the structure of included environments:

```suggestion
        f"""\
spack:
  include:
  - path: {path1}
    when: {include_condition}
  - path: {path2}
    when: {include_condition}
"""
```

alalazo commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -4635,3 +4633,226 @@ def test_non_str_repos(installed_environment):
+    )
+
+    for spec in specs:
+        assert Spec(spec) in e.user_specs
```
`SpecList` objects are currently unwieldy for these kind of checks, but I would do:
```python
assert [Spec(s) for s in specs] == e.user_specs.specs
```
to be more precise on what we expect.

alalazo commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -4635,3 +4633,226 @@ def test_non_str_repos(installed_environment):
+    includes another environment with the same definition but different spec(s),
+    results in user specs from both files."""
+    spack_yaml = """\
+spack:
+  definitions:
+  - packages: ["libelf"]
+  specs:
+  - $packages
+"""
```
In general would we be able to reuse `definitions:` from included environments?

alalazo commented on lib/spack/spack/test/util/remote_file_cache.py:
Changes in this file seem a good refactor that can be extracted, together with changes to `remote_file_cache.py`

alalazo commented on lib/spack/spack/test/config.py:
```diff
@@ -1612,8 +1613,7 @@ def test_config_include_similar_name(tmp_path: pathlib.Path):
     assert len(config.matching_scopes("^test$")) == 1
-    assert len(config.matching_scopes("^test:a/config$")) == 1
-    assert len(config.matching_scopes("^test:b/config$")) == 1
+    assert len(config.matching_scopes("^test:")) == 2
```
Why was this changed?

tldahlgren replied:
Nothing (anymore).

alalazo commented on lib/spack/spack/config.py:
```diff
@@ -1015,6 +1098,15 @@ def _scope(
 
+        # Ignore included environment lockfiles since they are not configuration
+        # files but are leveraging the common ``include`` configuration heading.
+        if path and os.path.basename(path) == "spack.lock":
```
This couples configuration and environment details by using `spack.lock`. Wondering if we can avoid this.

alalazo commented on lib/spack/spack/config.py:
```diff
@@ -1034,6 +1126,11 @@ def _scope(
+            # Paths create excessively long configuration names made worse with
+            # nested includes, so translate included path names to a hash.
+            if os.sep in included_name:
+                included_name = hashlib.sha1(included_name.encode("utf-8")).hexdigest()[:7]
```
Ok, I see why we need to change that assertion in the config tests. Still, this makes it hard to refer to scopes or construct their names given the path. 🤔

alalazo commented on lib/spack/spack/test/config.py:
```diff
@@ -1825,11 +1843,74 @@ def _checkout(*args, **kwargs):
+    """Confirm the default selection of paths."""
+    monkeypatch.setattr(spack.paths, "user_cache_path", str(tmp_path))
+
+    class MockIncludeGit(spack.util.executable.Executable):
```
We can probably extract this into a fixture that can be reused. I'll check how easy that is.

alalazo reviewed (commented):
Comments after a first read of the tests.

tldahlgren committed:
test_included_path_unwritable_dest: skip if root

Signed-off-by: tldahlgren <dahlgren1@llnl.gov>
Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

tldahlgren committed:
test_included_path_unwritable_dest: skip if root

Signed-off-by: tldahlgren <dahlgren1@llnl.gov>

tldahlgren committed:
Post-rebase cleanup of config.py

Signed-off-by: tldahlgren <dahlgren1@llnl.gov>
Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

tldahlgren committed:
Post-rebase cleanup of config.py

Signed-off-by: tldahlgren <dahlgren1@llnl.gov>

tldahlgren committed:
Update windows include tests sha256s

Signed-off-by: tldahlgren <dahlgren1@llnl.gov>
Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

tldahlgren committed:
Update windows include tests sha256s

Signed-off-by: tldahlgren <dahlgren1@llnl.gov>

tldahlgren committed:
Fix unwritable tempdir failure on rhel8

Signed-off-by: tldahlgren <dahlgren1@llnl.gov>

tldahlgren committed:
Fix unwritable tempdir failure on rhel8

Signed-off-by: tldahlgren <dahlgren1@llnl.gov>
Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

tldahlgren committed:
environments.rst: more rebase cleanup

Signed-off-by: tldahlgren <dahlgren1@llnl.gov>

tldahlgren committed:
environments.rst: more rebase cleanup

Signed-off-by: tldahlgren <dahlgren1@llnl.gov>
Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

tldahlgren committed:
Remove included_env_config and address selected style feedback

Signed-off-by: tldahlgren <dahlgren1@llnl.gov>
Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

tldahlgren committed:
Remove included_env_config and address selected style feedback

Signed-off-by: tldahlgren <dahlgren1@llnl.gov>

tldahlgren committed:
Restore test_config_include_similar_name checks

Signed-off-by: tldahlgren <dahlgren1@llnl.gov>
Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

tldahlgren committed:
Restore test_config_include_similar_name checks

Signed-off-by: tldahlgren <dahlgren1@llnl.gov>

tldahlgren committed:
specs.yaml: added schema support but not use

Signed-off-by: tldahlgren <dahlgren1@llnl.gov>

tldahlgren committed:
specs.yaml: added schema support but not use

Signed-off-by: tldahlgren <dahlgren1@llnl.gov>
Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
cleanup schema

Signed-off-by: Gregory Becker <becker33@llnl.gov>
Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
wip

Signed-off-by: Gregory Becker <becker33@llnl.gov>
Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
typo fix

Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
fixup

Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
refactor config included scopes to be easier to use in environments

Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
hotfix for schema while working on other code

Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
refactor usage of SpecListParser

Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
Environment._read: simplify codepaths

Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
environment concrete includes: refactor to pull from config system

Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
EnvironmentManifestFile: refactor to separate __init__ from reading speclists

Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
wrap calls to config in manifest.use_config()

Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
revert changes to relative include rewriting on env create

Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
fix reference-before-set bug in error handling code

Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
remove vestigial code from a previous attempt

Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
fix test for new config behavior

Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
fix type hint

Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
remove vestigial method

Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
modify specs inside config

Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
spec schema: deduplicate group_name_and_deps

Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
EnvironmentManifestFile.remove_default_view: inside of self.use_config():

Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
test/env.py: update tests for new calling conventions

Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
fixup env_include test for rebase since it was written

Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
revert bad test change

Signed-off-by: Gregory Becker <gbbmiler@gmail.com>
Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
style

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
EnvironmentManifestFile.use_config|prepare_config_scope: don't duplicate pushing scopes

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
spack.config: remove vestigial get_includes method and inline callsite

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
remove vestigial handling for nested configs

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
fixup after rebase

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
refactor schema to support merged config inside specs config

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
remove TODOs that are already accounted for

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
update commands to use env config over manifest

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
use config interface for definitions

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
prohibit including from manifest and lockfile of same environment

We don't know what this will mean semantically yet, so for now we
prohibit it so that we aren't stuck with a bad choice later.

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
cmd/uninstall: do not fail due to invalid environment elsewhere

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
fix test that relies on user environment root

binary distribution test relied implicitly on environment root
through uninstall command

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
style

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
more style

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
debug remote failure

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
more debugging

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
style

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
more fixing up debugging style

Signed-off-by: Gregory Becker <becker33@llnl.gov>

tgamblin commented on lib/spack/docs/include_yaml.rst:
```diff
@@ -75,52 +89,63 @@ The contents of the downloaded file are read and included in Spack's configurati
-And we want the configuration scope name to start ``common``.
-We could then configure the include in, for example, the user scope include file (i.e., ``$HOME/.spack/include.yaml`` by default), as follows::
+And we want the configuration scope name to start ``common`` and the files to be cached under ``$HOME/my_usc_includes``.
+We would then configure the ``include.yaml`` file as follows::
```
needs rephrasing

tgamblin commented on lib/spack/docs/include_yaml.rst:
```diff
@@ -75,52 +89,63 @@ The contents of the downloaded file are read and included in Spack's configurati
 .. note::
 
-   The git URL could be specified through an environment variable (e.g., ``$MY_USC_CONFIG_URL``).
+   The git URL could have been specified through an environment variable (e.g., ``$MY_USC_CONFIG_URL``).
```
```suggestion
   The git URL could also be specified through an environment variable (e.g., ``$MY_USC_CONFIG_URL``).
```

tgamblin commented on lib/spack/spack/config.py:
```diff
@@ -156,25 +155,67 @@ def __init__(self, name: str, included: bool = False) -> None:
-            if includes:
-                include_paths = [included_path(data) for data in includes["include"]]
-                included_scopes = chain(*[include.scopes(self) for include in include_paths])
+        assert isinstance(self._included_scopes, list)
```
how does this come up if the schema validates properly?

becker33 replied:
The assertion is to satisfy mypy.

The type is `Optional[List["ConfigScope"]]`, so before we can treat it as `List["ConfigScope"]` we have to have an assert to let mypy know we can safely do so.

tgamblin commented on lib/spack/spack/config.py:
```diff
@@ -156,25 +155,67 @@ def __init__(self, name: str, included: bool = False) -> None:
+                            ):
+                                msg = (
+                                    "Cannot include abstract and concrete from the same "
+                                    f"environment.\n    Environment: {parent}"
```
needs a better error message

tgamblin commented on lib/spack/spack/config.py:
```diff
@@ -156,25 +155,67 @@ def __init__(self, name: str, included: bool = False) -> None:
+            included_scopes = list(chain(*[include.scopes(self) for include in include_paths]))
+
+            for included_scope in included_scopes:
+                # If it's a lockfile, sort it accordingly
```
what does "accordingly" mean?  say what it actually does

tgamblin commented on lib/spack/spack/config.py:
```diff
@@ -1260,6 +1344,15 @@ def paths(self) -> List[str]:
         raise NotImplementedError("must be implemented in derived classes")
 
+    def to_dict(self) -> Dict[str, Any]:
+        """Return the minimal configuration information in dictionary form."""
```
what does minimal mean? why is it minimal?

becker33 replied:
No idea what that meant, fixing it.

tgamblin commented on lib/spack/spack/config.py:
```diff
@@ -1337,6 +1439,15 @@ def paths(self) -> List[str]:
         return [self.path]
 
+    def to_dict(self) -> Dict[str, Any]:
+        """Return minimal the configuration information in dictionary form."""
```
needs a clearer docstring -- what's minimal? why?

tgamblin commented on lib/spack/spack/config.py:
```diff
@@ -1486,8 +1598,37 @@ def scopes(self, parent_scope: ConfigScope) -> List[ConfigScope]:
+        manifest_filename = "spack.yaml"
+        with filesystem.working_dir(self.destination, create=False):
+            yaml_files = glob.glob("*.yaml")
+            # An environment manifest file takes precedence over any other
```
is included by default? what does "takes precedence" mean?

becker33 replied:
If you give a directory containing both a `spack.yaml` and other config files, it only includes the `spack.yaml` (we assume the other config files are included from the environment). If you delete the `spack.yaml` and include the directory, it will include it as a directory scope with all the normal files inside it.

tgamblin commented on lib/spack/spack/environment/environment.py:
```diff
@@ -1218,69 +1221,42 @@ def _load_concrete_include_data(self):
-        self._process_included_lockfiles()
-
+    # TODO/50207: Properly resolve the conflicts related to adding included
+    # TODO/50207: user specs.
```
I would not object if "user specs" became "roots" like it is in the docs but doesn't have to happen right now

tgamblin reviewed (changes requested):
some initial feedback -- looks mostly good; I don't see anything majorly wrong here, but need to review further.

kwryankrattiger commented on lib/spack/docs/include_yaml.rst:
```diff
@@ -75,52 +89,63 @@ The contents of the downloaded file are read and included in Spack's configurati
 
    include:
-   - name: common
+   - name: USC  # optional
```
I am confused why this changed. Note, none of the other references to this name have changed.

becker33 replied:
I think the primary motivation was to use an existing directory in the `github.com/spack/spack-configs` repo. I've updated it to match everywhere else in the description.

kwryankrattiger commented on lib/spack/docs/include_yaml.rst:
```diff
@@ -75,52 +89,63 @@ The contents of the downloaded file are read and included in Spack's configurati
+If the condition is satisfied, then the ``main`` branch of the repository will be cloned when the configuration scopes are initially created.
+The resulting repository can be found under ``$HOME/my_usc_includes``.
+
+Once cloned, the settings for the two files under the ``$HOME/my_usc_includes/USC/config`` subdirectory will be integrated into Spack's configuration.
```
```suggestion
Once cloned, the settings for the two files listed under ``paths`` will be integrated into Spack's configuration.
```

kwryankrattiger replied:
Not sure if this is better or not, but it is easier for me the parse back to the config field vs the system path

kwryankrattiger commented on lib/spack/docs/include_yaml.rst:
```diff
@@ -75,52 +89,63 @@ The contents of the downloaded file are read and included in Spack's configurati
    _builtin
 
 Since there are two unique paths, each results in a separate configuration scope.
 If only the ``USC/config`` directory was listed under ``paths``, then there would be only one configuration scope, named ``USC``, and the configuration settings from all of the configuration files within that directory would be integrated.
```
```suggestion
If only the ``USC/config`` directory was listed under ``paths``, then there would be only one configuration scope, named ``common``, and the configuration settings from all of the configuration files within that directory would be integrated.
```

This just needs to be consistent everywhere, I am in different to the name

kwryankrattiger commented on lib/spack/docs/include_yaml.rst:
```diff
@@ -75,52 +89,63 @@ The contents of the downloaded file are read and included in Spack's configurati
 For example, suppose we only want to include the ``config.yaml`` and ``packages.yaml`` files from the `spack/spack-configs <https://github.com/spack/spack-configs>`_ repository's ``USC/config`` directory when using the ``centos7`` operating system.
-And we want the configuration scope name to start ``common``.
-We could then configure the include in, for example, the user scope include file (i.e., ``$HOME/.spack/include.yaml`` by default), as follows::
+And we want the configuration scope name to start ``common`` and the files to be cached under ``$HOME/my_usc_includes``.
```
```suggestion
And we want the configuration scope name to start ``USC`` and the files to be cached under ``$HOME/my_usc_includes``.
```
Just needs to be consistent, `USC` or `common` are both fine.

kwryankrattiger commented on lib/spack/spack/schema/definitions.py:
```diff
@@ -10,7 +10,36 @@
 from typing import Any, Dict
 
-from .spec_list import spec_list_schema
+matrix_properties = {
```
This seems unrelated, should this be another PR?

becker33 committed:
address review on include docs

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
address review on comments/docstrings/messages in config.py

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
Remove TODO now that mixing manifest and lockfile is prohibited

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
very minor cleanup of env update

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
docs: clarify env installation life-cycle

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
simplify obfuscated callpath

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
more remote test failure debugging

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
remove debug print statements

Signed-off-by: Gregory Becker <becker33@llnl.gov>

becker33 committed:
remove vestigial import

Signed-off-by: Gregory Becker <becker33@llnl.gov>

kwryankrattiger commented on lib/spack/spack/cmd/ci.py:
```diff
@@ -16,7 +16,6 @@
 import spack.cmd.buildcache as buildcache
 import spack.cmd.common.arguments
 import spack.config as cfg
-import spack.environment as ev
```
I think all of the changes here could be another PR.

kwryankrattiger commented on lib/spack/spack/cmd/uninstall.py:
```diff
@@ -165,7 +166,16 @@ def dependent_environments(
     # Mapping from Environment -> non-zero list of specs contained in it.
     other_envs_to_specs: Dict[ev.Environment, List[spack.spec.Spec]] = {}
-    for other_env in (ev.Environment(ev.root(name)) for name in env_names):
+
```
I think this could be another separate PR.

kwryankrattiger commented on lib/spack/spack/environment/environment.py:
```diff
@@ -3103,17 +3077,37 @@ def _ensure_env_dir():
     shutil.copy(envfile, target_manifest)
 
-    # Copy relative path includes that live inside the environment dir
+    # Rewrite the relative path in the include path data
```
Make this a docstring

kwryankrattiger commented on lib/spack/spack/environment/environment.py:
```diff
@@ -3103,17 +3077,37 @@ def _ensure_env_dir():
         return
 
-    # TODO: make this recursive
+    # TODO GBB: Use the config system for this
```
```suggestion
    # TODO: Use the config system for this
```
I know who wrote this and had to think for a second who GBB was. Also, The blame here should be enough, I am not the biggest fan of dropping user names/aliases in TODO comments

kwryankrattiger commented on lib/spack/spack/environment/environment.py:
```diff
@@ -3103,17 +3077,37 @@ def _ensure_env_dir():
+    included_paths = [
+        include for include in included_paths if isinstance(include, spack.config.IncludePath)
+    ]
+    for idx, include in enumerate(included_paths):
```
Not seeing where `idx` is used

```suggestion
    for include in included_paths:
```

kwryankrattiger commented on lib/spack/spack/environment/environment.py:
```diff
@@ -3307,23 +3302,28 @@ def add_user_spec(self, user_spec: str, *, group: Optional[str] = None) -> None:
             group: group where the spec should be added. If None, the default group is used.
         """
-        group = group or DEFAULT_USER_SPEC_GROUP
+        with self.use_config():
```
Could these group config changes be another PR? I am not seeing where this is related in included environments.
